apis/v1alpha1/bucket_types.go

```diff
@@ -0,0 +1,161 @@
+/*
+Copyright 2021 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"reflect"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// BucketParameters are the configurable fields of a Bucket.
+type BucketParameters struct {
+	Description *string `json:"description,omitempty"`
+
+	// OrgID is the ID of the org this Bucket will be a member of.
+	// Either OrgID or OrgIDRef or OrgIDSelector has to be given during creation.
+	// +crossplane:generate:reference:type=Organization
+	// +crossplane:generate:reference:extractor=OrganizationID()
+	OrgID *string `json:"orgID,omitempty"`
+
+	// OrgIDRef references an Organization to retrieve its ID to populate OrgID.
+	// +optional
+	// +immutable
+	OrgIDRef *xpv1.Reference `json:"orgIDRef,omitempty"`
+
+	// OrgIDSelector selects a reference to an Organization to populate OrgIDRef.
+	// +optional
+	OrgIDSelector *xpv1.Selector `json:"orgIDSelector,omitempty"`
+
+	RP *string `json:"rp,omitempty"`
+
+	// Rules to expire or retain data. No rules means data never expires.
+	RetentionRules []RetentionRule `json:"retentionRules"`
+
+	// +kubebuilder:validation:Enum=implicit;explicit
+	SchemaType string `json:"schemaType,omitempty"`
+}
+
+// RetentionRule defines model for RetentionRule.
+type RetentionRule struct {
+	// Duration in seconds for how long data will be kept in the database. 0 means infinite.
+	EverySeconds int64 `json:"everySeconds"`
+
+	// Shard duration measured in seconds.
+	ShardGroupDurationSeconds *int64 `json:"shardGroupDurationSeconds,omitempty"`
+
+	// +kubebuilder:default=expire
```

Comment:
Is available types set somewhere, could we use enum?

Reply:
There is an enum but I don't think we can use them in the comment here and that's the only place we have these.